Fail the build if a referenced init script does not exist

patrikerdes opened the pull request:
If any of the init scripts specified on the command line does not exist
or is not a file, the build will fail.

Issue: #4672

Signed-off-by: Patrik Erdes <patrik@erdes.se>

### Context
I thought I could attempt to fix #4672. It follows the existing pattern for checking if files exist, both in the implementation and in the test.

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/.github/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#git-commit---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

lptr commented on subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/StartParamsValidatingActionExecuter.java:
```diff
@@ -60,6 +62,16 @@ public Object execute(BuildAction action, BuildRequestContext requestContext, Bu
                 throw new IllegalArgumentException(String.format("The specified settings file '%s' is not a file.", startParameter.getSettingsFile()));
             }
         }
+        if (startParameter.getInitScripts().size() > 0) {
```
This `if` is unnecessary.

patrikerdes replied:
Agreed, removed in 2a249918cf7e3bf430563cca65b908c3b5305eb1

wolfs commented on subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/StartParamsValidatingActionExecuter.java:
```diff
@@ -60,6 +62,16 @@ public Object execute(BuildAction action, BuildRequestContext requestContext, Bu
         }
+        if (startParameter.getInitScripts().size() > 0) {
+            for (File initScript : startParameter.getInitScripts()) {
+                if (!initScript.isFile()) {
```
It may be worth extracting this logic now into a method, given that we use it in three places now (`buildFile`, `settingFile`, `initScripts`).

patrikerdes replied:
I agree, fixed in 5ac77234718f4a3e4b077ab04f86db9f83ef04cc

wolfs replied:
👍

wolfs commented on subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/ProjectLoadingIntegrationTest.java:
```diff
@@ -150,6 +151,20 @@ public void buildFailsWhenSpecifiedSettingsFileIsNotAFile() {
+    public void buildFailsWhenSpecifiedInitScriptIsNotAFile() {
+        TestFile file = testFile("unknown");
+
+        ExecutionFailure result = inTestDirectory().usingInitScript(file).runWithFailure();
```
I guess we should have some tests where we have more than one init script argument on the command line, too. WDYT?

patrikerdes replied:
Sure, could be good to have. Does this look like what you had in mind? e28340725ead1e0f223f4327ef03121ac95a085d

wolfs replied:
👍

patrikerdes commented:
@lptr I agree that there is a risk of this breaking builds for people. How big that risk is, and what precautions are needed, I think all of you are better at judging than me.

I gave it a shot to update the release notes and docs: I rebased the branch on the master branch to get the current release notes, and added a note under "Potential breaking changes". I also updated the documentation page for init scripts.

lptr commented:
Oops, sorry, I missed that.

lptr commented:
@patrikerdes can you please sign off on your commits? You can use `git rebase --signoff` for that: https://git-scm.com/docs/git-rebase#git-rebase---signoff

After that I think we are good to merge.

lptr commented:
Branch build: https://builds.gradle.org/viewQueued.html?itemId=12246985

patrikerdes commented:
No you don't need to be sorry, you did not miss anything. I had not yet done the release notes and docs when you wrote your comment :)

Ok, I just did `git rebase --signoff` and force pushed.

patrikerdes commented:
But now it complains about me signing off the commit by Marc Philipp. Not sure how to fix that...

patrikerdes commented:
Regarding my previous comment about the DCO check failing: I assumed that the individual commits were not that interesting, so I just squashed it all together into one big commit. So now all checks pass.

lptr commented:
Branch build: https://builds.gradle.org/viewQueued.html?itemId=12267951

lptr commented:
Duh, will need to re-merge this because of CI troubles. Will try to do that today.

lptr commented:
Retry merge as #5078.

lptr commented:
Merged. This will be delivered as part of Gradle 4.8. Thanks @patrikerdes for your contribution!

lptr commented:
And of course the first build this broke was ours:

```text
* What went wrong:
The specified initialization script '/home/tcagent1/agent/work/668602365d1521fc/gradle/gradle/init-scripts/build-scan.init.gradle.kts' does not exist.
```

https://builds.gradle.org/viewLog.html?buildId=12367636&tab=buildResultsDiv&buildTypeId=Gradle_Check_SanityCheck
